hyperspy/_signals/lazy.py

```diff
@@ -521,87 +556,98 @@ def _map_all(self, function, inplace=True, **kwargs):
 
     def _map_iterate(self,
                      function,
-                     iterating_kwargs=(),
+                     iterating_kwargs={},
```

Comment:
Dicts (or lists or tuples) should not be used as default values for parameters like this, as it can lead to really strange bugs due to how python handles these things: https://docs.python-guide.org/writing/gotchas/

Change it to `iterating_kwargs=None` and further down:
```python
if iterating_kwargs is None:
    iterating_kwargs = None
```

Reply:
To clarify: the issue you are referring to is related to passing mutable object to functions and that mutable will be changed by function? The _pass by reference_ versus _pass by value_ story? `list` and `dict` are indeed mutable and will be indeed changed by the functions. In case of immutable (`tuple`, `str`, `int`, etc.) object, a copy will be made when it needs to be changed - because it is immutable.

The fact that `iterating_kwargs` is dict is what is concerning, because it this is a dictionary with immutable object as values, there is no issue. One possible concern is that `mutable` object are passed to the `map` method and modified by the `function` passed to `map`.

I don't see how setting `iterating_kwargs=None` will fix this issue and the only way to make sure mutable input object are not changed when calling `map` is to create a copy of the mutable objects in one way or another. If this is done by `map` or the responsibility of the user, I would say that we should leave this consideration for #2669, as for other improvements suggestion made in this PR!

Reply:
No, what I'm referring to is using mutable _default_ parameters, which can lead to some really, really strange results.

```python
def test_add_to_dict_bad(key, value, a={}):
    a[key] = value
    return a

def test_add_to_dict_correct(key, value, a=None):
    if a is None:
        a = {}
    a[key] = value
    return a

print(test_add_to_dict_bad("key", "value")) # {'key': 'value'}
print(test_add_to_dict_bad("key2", "value2")) # {'key': 'value', 'key2': 'value2'}

print(test_add_to_dict_correct("key", "value")) # {'key': 'value'}
print(test_add_to_dict_correct("key2", "value2")) # {'key2': 'value2'}
```

Essentially, if `iterating_kwargs={}` is used, the `iterating_kwargs` you used when you previously called the function will still be there. From https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments:

> Python’s default arguments are evaluated once when the function is defined, not each time the function is called (like it is in say, Ruby). This means that if you use a mutable default argument and mutate it, you will and have mutated that object for all future calls to the function as well.

This is one of the common "gotchas" in python, so I prefer using the `parameter=None`, followed by `if parameter is None: ...` trick, for list, dicts, (even tuples).

Reply:
I see indeed, this is different, but this issue doesn't occur here, because no key is added to `iterating_kwargs`. Adapting your example above, this is equivalent to:
```python
def test_add_to_dict_not_that_bad(key, value, a={}):
    for key in a.keys():
        a[key] = value
    
    return a

print(test_add_to_dict_not_that_bad("key", "value")) # {'key': 'value'}
print(test_add_to_dict_not_that_bad("key2", "value2")) # {'key2': 'value2'}

print(test_add_to_dict_not_that_bad("key", "value", {'my_key':1})) # {'my_key1': 'value'}
print(test_add_to_dict_not_that_bad("key", "value", {'my_key2':2})) # {'my_key2': 'value'}
```

The `iterating_kwargs` is also created before [being passed as argument](https://github.com/CSSFrancis/hyperspy/blob/5defd64b77f818c9478c6f790f76d16918c8428e/hyperspy/signal.py#L4501-L4512) and users are not expected to call `_map_iterate` directly.